update dill docs

tinnus-napbus opened the pull request:
- bring api ref up to date
- add scry ref
- add data types ref
- rewrite overview

Overview is very brief, maybe want to expand - maybe see what @Fang- thinks, wasn't sure exactly what should be described in the overview.

resolves #1073

tinnus-napbus committed:
update dill docs
- bring api ref up to date
- add scry ref
- add data types ref
- rewrite overview

drbeefsupreme commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,15 @@ weight = 1
-simply pushed onto the buffer and blitted to the screen, but some
-characters, including control-modified keys, arrow keys, etc. require
-special handling. Most of the readline functionality is in `%dill`.
+Unix sends keyboard events and the like to Dill, and Dill produces terminal output. Other vanes interact with Dill by passing it [task](/docs/arvo/dill/tasks)s to do things like print error messages. For userspace applications on the other hand, one wouldn't typically deal with Dill directly. Rather, the `%hood` agent's `drum` module acts as an intermediary session handler for CLI applications in ...
```
```suggestion
Unix sends keyboard events and the like to Dill, and Dill produces terminal output. Other vanes interact with Dill by passing it [tasks](/docs/arvo/dill/tasks)s to do things like print error messages. For userspace applications on the other hand, one wouldn't typically deal with Dill directly. Rather, the `%hood` agent's `drum` module acts as an intermediary session handler for CLI applications in userspace. To help write CLI apps there's two convenient libraries, `shoe` and `sole`, and a [tutorial](/docs/hoon/guides/cli-tutorial) for how to use them.
```

drbeefsupreme commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,15 @@ weight = 1
+
+[Scry Reference](/docs/arvo/dill/scry) - The scry endpoints of Dill.
+
+[Data Types](/docs/arvo/dill/data-types) - Reference documentation of the data types used by Jael.
```
```suggestion
[Data Types](/docs/arvo/dill/data-types) - Reference documentation of the data types used by Dill.
```

drbeefsupreme commented on content/docs/arvo/dill/tasks.md:
````diff
@@ -0,0 +1,339 @@
+[%boot lit=? p=*]
+```
+
+This `task` is used only once, when Arvo first enters the [adult stage](/docs/arvo/overview#structural-interface-core). Dill is technically the first vane to be activated, via the `%boot` `task`, which then sends Jael (considered the "true" first vane) the `%dawn` or `%fake` `task` wrapped in the `%boot`. Jael then goes on to call `%init` `task`s for other vanes (including Dill).
````
```suggestion
This `task` is used only once, when Arvo first enters the [adult stage](/docs/arvo/overview#structural-interface-core). Dill is technically the first vane to be activated, via the `%boot` `task`, which then sends Jael (considered the "true" first vane) the `%dawn` or `%fake` `task` wrapped in the `%boot` `task`. Jael then goes on to call `%init` `task`s for other vanes (including Dill).
```

drbeefsupreme commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,339 @@
+
+#### Returns
+
+Dill returns no `gift` in response to a `%boot` `task`, but it will pass the wrapped `%dawn` or `%fake` `task` to Jael.
```
```suggestion
Dill returns no `gift` in response to a `%boot` `task`, but it will `%pass` the wrapped `%dawn` or `%fake` `task` to Jael.
```

drbeefsupreme commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,339 @@
+
+Produce memory report.
+
+When Dill receives a `%heft` `task` it passes arvo a `%whey` `waif` to obtain a memory report and prints it to the terminal.
```
```suggestion
When Dill receives a `%heft` `task` it passes Arvo a `%whey` `waif` to obtain a memory report and prints it to the terminal.
```

drbeefsupreme reviewed (changes requested):
LGTM but I think there's one short thing you could add to make the relationship between Dill, the runtime, and userspace a bit more clear, perhaps on the Overview page. Give a quick explanation of what happens when you turn on `|verb` in dojo and then press a key, say `a`:

```
["" %unix %belt /d/term/1 ~2021.9.2..16.40.28..1eaf]
["|" %pass [%dill %g] [[%deal [~zod ~zod] %hood %poke] /] ~[//term/1]]
["||" %give %gall [%unto %poke-ack] i=/dill t=~[//term/1]]
["||" %give %gall [%unto %fact] i=/dill t=~[//term/1]]
["|||" %give %dill %blit i=/gall/use/herm/0w3.gsXou/~zod/view/ t=~[/dill //term/1]]
["|||" %give %dill %blit i=/gall/use/herm/0w3.gsXou/~zod/view/ t=~[/dill //term/1]]
[ "||"
  %pass
  [%gall %g]
  [[%deal [~zod ~zod] %dojo %poke] /use/hood/0w3.gsXou/out/~zod/dojo/drum/phat/~zod/dojo]
  ~[/dill //term/1]
]
[ "|||"
  %give
  %gall
  [%unto %poke-ack]
  i=/gall/use/hood/0w3.gsXou/out/~zod/dojo/drum/phat/~zod/dojo
  t=~[/dill //term/1]
]
```

Maybe this could go on an examples pages instead. I think you could be pretty brief here - just saying what data structures are being included in the `task`s, and then linking to `/arvo/tutorials/move-trace.md` for a more thorough explanation of how to understand the output.

Looking at it now, the move trace page may need to be updated. I am seeing `%blit`s in the output now when in the past verbose mode omitted them and so I added them in by hand when writing the move trace page.

tinnus-napbus committed:
dill: fix typo "jael" -> "dill

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
dill: fix typo - missing "task"

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
dill:tasks - fix typo - add backticks to pass

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
dill:tasks - capitalise "Arvo"

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus commented:
ok hmm I realise I misunderstood part of dill, I'm gonna clarify a couple of things and rework some of this a bit along with a fleshed out overview and so forth

tinnus-napbus commented:
also it seems the `%blit`s to unix are still suppressed. What you're seeing are only the `%blit`s to `%herm` which subscribes to `%blit`s and translates them to json for the web terminal afaict

tinnus-napbus committed:
dill: minor grammar fix

tinnus-napbus committed:
dill: expand overview, add diagram

tinnus-napbus committed:
dill: minor updates to tasks

tinnus-napbus commented:
ok I've expanded the overview page with a better explanation. Rather than a move-trace I've included a diagram. Currently it links to https://m.tinnus-napbus.xyz/pub/dill-userspace.svg - @drbeefsupreme if you have access to the urbit.org s3 instance, could you upload that there instead?

@Fang- I've made the changes I mentioned, you can review now

Fang- commented on content/docs/arvo/dill/_index.md:
```diff
@@ -5,3 +5,19 @@ sort_by = "weight"
+
+[Data Types](/docs/arvo/dill/data-types)
+
+Reference documentation of the data types used by Jael.
```
```suggestion
Reference documentation of the data types used by Dill.
```

Fang- commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+
+The terminal driver vane.
+
+Keyboard events and the like from Unix are received by Dill as [%belt](/docs/arvo/dill/tasks#belt) `task`s, and Dill sends [%blit](/docs/arvo/dill/tasks#blit) `gift`s back to the runtime to be displayed in the Unix terminal. The manner of interacting with Dill differs depending on whether you're in userspace or kernelspace, as we'll explore below.
```
I see no blit section on the tasks page, perhaps you meant:

```suggestion
Keyboard events and the like from Unix are received by Dill as [%belt](/docs/arvo/dill/tasks#belt) `task`s, and Dill sends [%blit](/docs/arvo/dill/data-types#blit) `gift`s back to the runtime to be displayed in the Unix terminal. The manner of interacting with Dill differs depending on whether you're in userspace or kernelspace, as we'll explore below.
```

Fang- replied:
There's also a small caveat here that "the Unix terminal" is really anything connected to Dill, which may be the runtime, a standalone binary that mimics the runtime's terminal behavior, a graphical client that renders the blits in a special way, etc.

Of course mentioning that may be a little pointless or unclear when no such things actually exist in the wild (does webterm count?), so do with this what you will.

Fang- commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+
+## Kernelspace
+
+For technical reasons, Dill performs a handful of system tasks related to booting a ship and some memory operations. Aside from those, other Vanes mostly just pass Dill [tasks](/docs/arvo/dill/tasks) to print error messages and the like in the terminal.
```
```suggestion
For technical reasons, Dill performs a handful of system tasks related to booting a ship and some memory operations. Aside from those, other Vanes mostly just pass Dill [tasks](/docs/arvo/dill/tasks) to print error messages and the like to the terminal.
```

This is a nit, but printing _to_ something feels better than printing _in_ something.

Fang- commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+
+## Userspace
+
+Unlike in kernalspace, userspace applications are unlikely to `%pass` Dill `task`s directly. Instead, Dill looks at things in terms of sessions. A session is a pipeline between a client and a handler, where:
```
```suggestion
Unlike in kernelspace, userspace applications are unlikely to `%pass` Dill `task`s directly. Instead, Dill looks at things in terms of sessions. A session is a pipeline between a client and a handler, where:
```

Fang- commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+- The client is an external input source and output sink; a terminal with with dimensions and so forth.
+- The handler is an application in Urbit that interprets input, maybe does something with it, maybe produces output to be displayed in the client, etc.
+
+There are plans for Dill to support multiple independent sessions (i.e. independent terminal clients displaying and processing different information at the same time), but at this stage Dill only supports a single session. The client for this default session is just the Unix terminal via the Urbit runtime. The handler for the default session is `drum`, a submodule of the the `%hood` agent.
```
```suggestion
There are plans for Dill to support multiple independent sessions (i.e. independent terminal clients displaying and processing different information at the same time), but at this stage Dill only supports a single session. The client for this default session is just the Unix terminal via the Urbit runtime. The handler for the default session is `drum`, a submodule of the `%hood` agent.
```

Also this is the only place you say "agent" instead of "app" or "application". Probable best to be consistent.

Technically the handler is `%hood`, which passes all terminal-related stuff on to `drum` for actual handling, but I'm not sure if the directionality presented here is that important.

Fang- commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+
+To give a basic idea of how keyboard events flow through these systems and produce terminal output, here's a diagram showing the messages in pseudo-hoon:
+
+![Dill userspace diagram](https://m.tinnus-napbus.xyz/pub/dill-userspace.svg)
```
Nice! This should/will be hosted on media.urbit.org, right?

Fang- commented on content/docs/arvo/dill/data-types.md:
```diff
@@ -0,0 +1,187 @@
+
+Terminal input.
+
+A `$belt` is passed to Dill in a [%belt](/docs/arvo/dill/tasks#belt) `task` by the runtime whenever there is input, such as a user typing in the console. This is the old `$belt` and is only used by the runtime, the new [$dill-belt](#dill-belt) is what is used between Dill and userspace.
```
> This is the old `$belt`

I know the comments in lull.hoon say "old" and "new", but this is a bit misleading, since these aren't actually separate versions of the api. I think they're meant as "old world" and "new world" respectively, as in talking to earth vs talking to mars.

The way I've chosen to describe these going forward is that `$belt` and `$blit` are _client_ input and output, whereas `$dill-belt` and `$dill-blit` are _arvo_ input and output, since these pairs comes from & go to the client and arvo respectively.

(This obviously also applies to the other mentions of old/new below.)

Fang- commented on content/docs/arvo/dill/data-types.md:
```diff
@@ -0,0 +1,187 @@
+- `%mod` - Modifier (Ctrl, Meta or Hyper) plus [key].
+- `%txt` - A series of characters
+- `%ctl` - Ctrl+[key]
+- `%met` - Meta+[key]
```
Worth noting that these are deprecated in favor of `%mod` and will be removed as part of the terminal stack refactor.

Fang- commented on content/docs/arvo/dill/data-types.md:
```diff
@@ -0,0 +1,187 @@
+- `%aro` - Arrow keys.
+- `%bac` - Backspace key.
+- `%del` - Delete key.
+- `%hit` - Mouse click - `r` is row and `c` is column.
```
Also notable, and perhaps less obvious: row and column values are zero-indexed, with `[0 0]` being the _bottom left_-most coordinate on the screen (as opposed to the xterm(/unix?) standard of basing this on the _top_ left). This is documented [here](https://github.com/urbit/urbit/blob/42e8158be9a53d907c5141188cbd3d6760111632/pkg/urbit/vere/io/term.c#L397) and in various commit messages, but for some reason not in lull.hoon. My bad!

Fang- commented on content/docs/arvo/dill/data-types.md:
```diff
@@ -0,0 +1,187 @@
+
+Terminal output.
+
+A `$blit` is given to the runtime by Dill in a [%blit](/docs/arvo/dill/tasks#blit) `task` when it wants to print some text, clear the screen, go _ding_ or what have you.
```
This is a gift, right? Also again, no dedicated blit section on the tasks page.

```suggestion
A `$blit` is given to the runtime by Dill in a [%blit](data-types#blit) `gift` when it wants to print some text, clear the screen, go _ding_ or what have you.
```

Fang- commented on content/docs/arvo/dill/data-types.md:
```diff
@@ -0,0 +1,187 @@
+- `%klr` - Set styled line, the `$stub` specifies the text and style.
+- `%lin` - Set current line, `p` contains the text.
+- `%mor` - Newline.
+- `%sag` - Save to jamfile, typically in `/[pier]/.urb/put/`. `p` is `/[optional path]/[filename]/[extension]`. For example, `/foo/bar` will save it in `/[pier]/.urb/put/foo.bar` and `/a/b/c/foo/bar` will save it in `/[pier]/.urb/put/a/b/c/foo.bar`. `q` is the `noun` to `jam` and save in the file.
```
Fun fact: even the filename here is optional. You can totally `.ext 'abc'` in dojo and the resulting `%sag` will happily write to `.urb/put.ext`!

Looking at the implementation for these "save to file" blits now, it seems they also accept the _empty_ path. This might be considered no bueno, and we may or may not want to consider changing the type to disallow the empty path, but right now nothing that ships with userspace lets you use it that way.

Fang- commented on content/docs/arvo/dill/data-types.md:
```diff
@@ -0,0 +1,187 @@
+- `%lin` - Set current line, `p` contains the text.
+- `%mor` - Newline.
+- `%sag` - Save to jamfile, typically in `/[pier]/.urb/put/`. `p` is `/[optional path]/[filename]/[extension]`. For example, `/foo/bar` will save it in `/[pier]/.urb/put/foo.bar` and `/a/b/c/foo/bar` will save it in `/[pier]/.urb/put/a/b/c/foo.bar`. `q` is the `noun` to `jam` and save in the file.
+- `%sav` - Save to file. Same behaviour as `%sag` except `q` is an `atom` rather than a `noun` and therefore doesn't need to be `jam`med.
```
The atom is written to disk as if it were the bytestring in the tail of an `octs`. That is, `%sav`ing the cord `'abcdef'`, whose `@ux` value is `0x6665.6463.6261`, results in a unix file whose hex dump renders as `61 62 63 64 65 66`.

Fang- commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,343 @@
+
+The [$blew](/docs/arvo/dill/data-types#blew) specifies the new dimensions.
+
+Dill will convert the `$blew` into a `%rez` [$dill-belt](/docs/arvo/dill/data-types#dill-belt) and `%poke` the session handler (typically `drum`) with it.
```
```suggestion
Dill will convert the `$blew` into a `%rez` [$dill-belt](/docs/arvo/dill/data-types#dill-belt) and `%poke`s the session handler (typically `drum`) with it.
```

Fang- commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,343 @@
+
+Unwatch a session to which you've previously subscribed with [%view](#view).
+
+The ability to specify a session is not yet implemented in Dill, so `session` is always `~` and it's always the default session.
```
```suggestion
The ability to specify a session is not yet implemented in Dill, so `session` is always `~`, the default session.
```

Fang- commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,343 @@
+
+Watch session.
+
+A `%view` `task` subscribes for a copy of all `%blit`s which Dill sends to the Unix terminal. This `task` is used by the `%herm` app so it can convert the `$blit`s to JSON and mirror the Unix terminal in the web terminal.
```
> subscribes for a copy of ...

Hmm, perhaps this is pedantic, but I wouldn't describe it this way. Rather, a `%view` by the runtime is sort of implicit in `%boot`. Sending a `%view` merely adds another subscriber to the same endpoint that the runtime/Unix terminal happens to be listening on. Herm then, too, is just subscribing to dill sessions, rather than the specific thing that the runtime is interested in. (Mirroring is presently indeed the result though.)

Fang- commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,343 @@
+
+A `%view` `task` subscribes for a copy of all `%blit`s which Dill sends to the Unix terminal. This `task` is used by the `%herm` app so it can convert the `$blit`s to JSON and mirror the Unix terminal in the web terminal.
+
+The ability to specify a session is not yet implemented in Dill, so `session` is always `~` and it's always the default session.
```
```suggestion
The ability to specify a session is not yet implemented in Dill, so `session` is always `~`, the default session.
```

Fang- commented on content/docs/arvo/dill/tasks.md:
```diff
@@ -0,0 +1,343 @@
+
+#### Returns
+
+Dill will `%give` a copy of all `%blit`s it sends to the Unix terminal. A `%blit` `gift` is:
```
See also the above comment.

Fang- commented on content/docs/arvo/dill/tasks.md:
````diff
@@ -0,0 +1,343 @@
+```
+
+`%trim` is a common vane `task` used to reduce memory usage. However, it does
+nothing for Dill.
````
May or may not be worth mentioning this is because dill keeps only very minimal and necessary state.

Fang- reviewed (changes requested):
Thank you as always!

Kind of amused at the amount of tasks documented here that are just straight-up not used. Looking forward to giving these all proper homes/burials.

The general phasing of "terminal input" rubs up against the vision of having this be modeled as broader "user input" instead. But then, that's just a _vision_, and not the reality we live in, so for technical/reference documentation this should be alright for the time being.

Similarly, as you're aware, these docs will need to be updated for the terminal refactor once that goes live (Soon™!), but we can cross that bridge when we get there. Changes shouldn't be _too_ big, in any case.

Mostly small corrections/suggestions below.

tinnus-napbus committed:
dill: typo, jael->dill

Co-authored-by: fang <github@fang.io>

tinnus-napbus committed:
dill: minor grammar, in->to

Co-authored-by: fang <github@fang.io>

tinnus-napbus committed:
dill: typo, kernalspace->kernelspace

Co-authored-by: fang <github@fang.io>

tinnus-napbus committed:
dill: grammar, poke->pokes

Co-authored-by: fang <github@fang.io>

tinnus-napbus committed:
dill: tasks - %flee - clarify default session wording

Co-authored-by: fang <github@fang.io>

tinnus-napbus committed:
dill: change diagram to media.urbit.org

tinnus-napbus committed:
dill: overview - fix %blit link

tinnus-napbus committed:
dill: overview - fix phrasing re: %hood

tinnus-napbus committed:
dill: clarify blit/belt new/old wording

tinnus-napbus committed:
dill: clarify %ctl/%met deprecated

tinnus-napbus committed:
dill: clarify %hit 0-indexed bottom-left

tinnus-napbus committed:
dill: fix %blit gift non-link

tinnus-napbus committed:
dill: expand %sav and %sag details

tinnus-napbus committed:
dill: reword %view task

tinnus-napbus committed:
dill: slight %view rewording

Co-authored-by: fang <github@fang.io>

tinnus-napbus committed:
dill: %view - returns reword

tinnus-napbus committed:
dill: clarify %trim re: dill's state

tinnus-napbus commented:
ok I've addressed the points @Fang- raised.

Also I now have access to the urbit s3 so I've uploaded the diagram and changed the link, so @drbeefsupreme you don't need to do that anymore

Fang- reviewed (approved):
All good changes, thank you!

drbeefsupreme commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+
+While `drum` talks with Dill in `$dill-belt`s and `$dill-blit`s, it talks to CLI apps with `$sole-action`s and `$sole-event`s, which are defined in the `sole` library. For more information on the `sole` library and the related `shoe` library, and for information on how to build CLI apps, you can refer to the [CLI app tutorial](/docs/hoon/guides/cli-tutorial).
+
+To give a basic idea of how keyboard events flow through these systems and produce terminal output, here's a diagram showing the messages in pseudo-hoon:
```
```suggestion
To give a basic idea of how keyboard events flow through these systems and produce terminal output, here's a diagram showing the messages in pseudo-Hoon:
```

drbeefsupreme commented on content/docs/arvo/dill/dill.md:
```diff
@@ -4,21 +4,36 @@ weight = 1
+To give a basic idea of how keyboard events flow through these systems and produce terminal output, here's a diagram showing the messages in pseudo-hoon:
+
+![Dill userspace diagram](https://media.urbit.org/docs/arvo/dill/dill-userspace.svg)
+
```
```suggestion
You can use a [move trace](/docs/arvo/tutorials/move-trace) to get a hands-on feel for this data flow.
```

drbeefsupreme reviewed (changes requested):
LGTM, the diagram is great!

To fix the build, merging master into the branch might work.

tinnus-napbus committed:
dill: minor grammar fix - capitalise Hoon

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
dill - add link to move trace guide

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
Merge branch 'urbit:master' into update-dill-docs

tinnus-napbus committed:
dill: fix move-trace remark layout

tinnus-napbus commented:
@drbeefsupreme cool yeah merging master seems to have fixed vercel's problem
